### PR TITLE
FIX set GLUM sampling to tolerance as only 1 iter is run for Lasso

### PR DESCRIPTION
**Pull request by tomMoral**

### CODE REVIEW

File: solvers/glum.py [hunk 1]
```diff
@@ -25,11 +28,12 @@ def set_objective(self, X, y, lmbd, fit_intercept):
...
             fit_intercept=fit_intercept,
             gradient_tol=1e-8,
+            max_iter=1
         )
         warnings.filterwarnings('ignore', category=ConvergenceWarning)
 
-    def run(self, n_iter):
-        self.model.max_iter = n_iter + 1
+    def run(self, tol):
+        self.model.gradient_tol = tol ** 2
```
Comment (tomMoral) [inline solvers/glum.py:36]: I am not sure this is the right scaling.
I would say we should remove the `**2` and scale this with the norm of the gradient in 0?

**Comment by mathurinm** [comment]
It seems glum is not designed (in its current state) for large scale Lasso:
- it support sparse data (contrary to what we thought) but instanciates a Hessian matrix so fails for finance. `rcv1` works on my machine, but swaps (`n_features=47_000`). See https://github.com/Quantco/glum/issues/485
- it seems that supporting this is on their mid-term agenda: https://github.com/Quantco/glum/issues/133   https://github.com/Quantco/glum/issues/211

**Comment by tbng** [comment]
I have the same memory issue for `glum` on `finance` dataset (with a Hessian matrix being initialized that leads to out of memory error). For the fix of the convergence of `glum` with simulated dataset, the advice from IRL discussion with @tomMoral and @agramfort would be to look at the `glum` source code and checking for right scaling, but my schedule is quite tight at the moment. I will try to get back to it as soon as I can.

**Comment by agramfort** [comment]
don't fix the problem, just tell the authors about it and tell them how to
replicate the issue

>

**Comment by mathurinm** [comment]
@agramfort  It is done, I am the author of the above issues on the glum repo.
As for figuring out how to use their various stopping criteria, I will look into it myself @tbng and keep you posted

**Comment by tbng** [comment]
> @agramfort It is done, I am the author of the above issues on the glum repo. As for figuring out how to use their various stopping criteria, I will look into it myself @tbng and keep you posted

Thanks a lot! Sorry it's really close to my thesis defense now so I really cannot bring much help at this moment.

**Comment by mathurinm** [comment]
My understanding is that glum normalizes even if fit_intercept=False and P1 = "identity". I manage to get it to work when I normalize and center X, otherwise their stopping criterion does not correspond to the one I compute (L1 norm of minimal  norm subgradient). I have pinged them on their repo.

In the meantime I managed to approximately reverse engineer their stopping criterion to produce the following figure:

image

I had to increaase patience to 5 otherwise GLUM was stopping too early, as it does not make progress from `tols[1]` to `tols[4]`[

File: solvers/glum.py [hunk 2]
```diff
@@ -14,23 +14,33 @@ class Solver(BaseSolver):
...
         )
         warnings.filterwarnings('ignore', category=ConvergenceWarning)
 
-    def run(self, n_iter):
-        self.model.max_iter = n_iter + 1
-        self.model.fit(self.X, self.y)
+    def run(self, tol):
+        # support sparse while avoiding MemoryError if n_features too large,
+        # as glum intanciates a Hessian
+        if self.X.shape[1] < 20_000:
```
Comment (agramfort) [inline solvers/glum.py:41]: why don't you use the skip method here?

Comment (mathurinm) [reply solvers/glum.py:41]: I was not aware of it. I'll add it !

**Comment by tomMoral** [comment]
Thx @mathurinm !! :)

**Pull request merged**